# Week4 — RAG 성능 개선 + 평가

지난 주 (week3) 의 외부 DB 위에 **표준 RAG (retrieve → generate)** 파이프라인을 띄우고, 평가용 질문 20개를 정의한 뒤, 두 가지 개선 실험을 정량 + 정성으로 비교합니다.

기술 스택은 우리 제품 **bsvibe-app** 의 실제 코드를 그대로 가져왔습니다.

목차

1. Baseline 구성
2. 평가 질문셋 (20문항)
3. Baseline 실제 동작 (Q → top-K → A) — top-K 전부 노출
4. Baseline 전체 결과 요약
5. 성능 개선 실험 (Baseline vs EXP-1 vs EXP-2)
6. 결론


## 1. Baseline 구성 (과제 B 항목)

| 항목 | 값 |
|---|---|
| Text Splitter | paragraph-boundary + max 420 chars (week2 chunking) |
| Chunk Size | 420 |
| Chunk Overlap | 0 |
| Embedding Model | `ollama/bge-m3` (1024d, BAAI multilingual) |
| Vector Store | `pgvector` (rag_chunks 테이블, `vector(1024)`, ivfflat lists=10) |
| Retriever | cosine similarity, top_k = **5** |
| Prompt | system 메시지 1줄 + concatenated context + user 질문 |
| LLM | `ollama/llama3.2:3b` (temperature 0.0, max_tokens 512) |


## 2. 질문셋 (20문항)

`data/eval/questions.jsonl` — 난이도별로 마킹.

- `simple_fact` (단순 사실 조회) — 10문항
- `multi_doc` (다중 문서 종합) — 6문항
- `compare` (조건 비교) — 4문항


In [ ]:
from pathlib import Path
import json
from collections import Counter

QS_PATH = Path.cwd().resolve()
while QS_PATH.name != "metacode_study" and QS_PATH != QS_PATH.parent:
    QS_PATH = QS_PATH.parent
QS_PATH = QS_PATH / "assignments/week4-rag-eval/data/eval/questions.jsonl"
questions = [json.loads(l) for l in QS_PATH.read_text(encoding="utf-8").splitlines() if l.strip()]
print(f"총 {len(questions)}문항")
print(f"난이도 분포: {dict(Counter(q['difficulty'] for q in questions))}")
print()
print("처음 5문항:")
for q in questions[:5]:
    print(f"  [{q['qid']:6}] [{q['difficulty']:11}] expected={q['expected_source_ids']}  Q: {q['question'][:50]}")


총 20문항
난이도 분포: {'simple_fact': 10, 'multi_doc': 6, 'compare': 4}

처음 5문항:
  [q-001 ] [simple_fact] expected=['bsvibe-src-001']  Q: BSVibe 모노레포는 어떤 구성으로 돌아가나요?
  [q-002 ] [simple_fact] expected=['bsvibe-src-002']  Q: workspace 단위로 vault와 retriever를 어떻게 묶나요?
  [q-003 ] [multi_doc  ] expected=['bsvibe-src-004', 'bsvibe-src-015']  Q: 이미 결정된 내용을 다음 run에서 재사용하려면 어디를 봐야 하나요?
  [q-004 ] [simple_fact] expected=['bsvibe-src-005']  Q: 파운더가 거절한 접근(부정 사례)은 어디에 저장되고 어떻게 검색되나요?
  [q-005 ] [simple_fact] expected=['bsvibe-src-006']  Q: canonical concept, decision, negative pattern을 한 번


## 3. Baseline 실제 동작 시연 (Q → top-K → A)

루브릭 ③ — DB 구축 직후 임의의 쿼리에 ``similarity_search_with_score`` 로 top-K + 점수 확인. **모든 결과에 source_id 옆에 title 표시** (week3 감점 사유 재발 방지).

대표 3문항을 실제 baseline pipeline 으로 돌린 출력입니다.


In [ ]:
from pathlib import Path
import asyncio, json, os, sys
from dotenv import load_dotenv
from sqlalchemy.ext.asyncio import create_async_engine

ROOT = Path.cwd().resolve()
while ROOT.name != "metacode_study" and ROOT != ROOT.parent:
    ROOT = ROOT.parent

sys.path.insert(0, str(ROOT / "assignments/week4-rag-eval/scripts"))
load_dotenv(ROOT / ".env")
os.environ.setdefault("OPENAI_API_KEY", "sk-noop")

from pipeline import BaselineRagPipeline

engine = create_async_engine("postgresql+asyncpg://rag:rag@localhost:5433/rag_week3")
pipe = BaselineRagPipeline(engine=engine, top_k=5)

PICKS = ["q-001", "q-010", "q-018"]
async def demo():
    for q in questions:
        if q["qid"] not in PICKS:
            continue
        ans = await pipe.answer(q["question"])
        expected = set(q["expected_source_ids"])
        print(f"[{q['qid']}] {q['question']}")
        print(f"  difficulty: {q['difficulty']} | expected: {q['expected_source_ids']}")
        print(f"  Retrieved top-5 (cosine sim):")
        for i, h in enumerate(ans.hits, 1):
            mark = " ✓" if h.source_id in expected else ""
            print(f"    {i}. sim={h.similarity:.3f}  {h.source_id:18}{mark}  {h.title if hasattr(h, 'title') else ''}")
        print(f"  ANSWER:")
        print("   ", ans.answer[:600])
        print()

asyncio.get_event_loop().run_until_complete(demo())


[q-001] BSVibe 모노레포는 어떤 구성으로 돌아가나요?
  difficulty: simple_fact | expected: ['bsvibe-src-001']
  Retrieved top-5 (cosine sim):
    1. sim=0.607  bsvibe-src-001     ✓  
    2. sim=0.441  bsvibe-src-030      
    3. sim=0.419  bsvibe-src-004      
    4. sim=0.385  bsvibe-src-006      
    5. sim=0.381  bsvibe-src-031      
  ANSWER:
    BSVibe 모노레포는 FastAPI 백엔드와 Next.js PWA를 포함하여 unified monorepo로 구축되어 있습니다. Docker Compose를 사용하여 프로젝트를 시작하고, 배포를 위해 compose.prod.yaml와 compose.yaml 두 가지 파일을 사용합니다.

compose.prod.yaml은 compose.yaml에 위함하는 추가적인 설정을 포함합니다. Specifically,它는 다음을 수행합니다:

* env-driven values를 사용하여 BSVIBE_DATABASE_URL, redis url, KMS key, 및 Supabase config를 대체합니다.
* public Postgres와 Redis 포트를 publishes하지 않습니다.
* BSVIFE_WORKER_MODE를 db_polling로 설정합니다.
* GIT_SHA를 fixed로 설정하여 프로젝트의 빌드 버전을 확인할 수 있습니다.

이러한 구성은 BSVibe 모노레포가 구축된 상태에서 배포 및 운영에 사용됩니다.

[q-010] 파운더가 잘못된 노드를 retract하면 어떤 흐름으로 처리되나요?
  difficulty: multi_doc | expected: ['bsvibe-src-016', 'bsvibe-src-022', 'bsvibe-src-024']
  Retr

## 4. Baseline 전체 결과 요약

20문항 전체에 대해 baseline 을 돌린 집계 — 아래 §5 에서 두 실험과 비교됩니다.

| 지표 | baseline |
|---|---|
| Hit@1 (top-1 안에 expected) | **95.0%** (19/20) |
| Hit@3 | 100.0% |
| Hit@5 | 100.0% |
| MRR | 0.967 |
| 평균 응답 시간 | 3.1s/질문 |
| 누적 wall-clock | 62.0s |

LLM-as-judge (gpt-4o-mini, 4 기준 0-3 점) 결과:

| faithfulness | relevance | completeness | citation | total /12 | normalized |
|---|---|---|---|---|---|
| 2.80 | 3.00 | 2.20 | 1.20 | **9.20** | **0.767** |

- 검색은 사실상 완벽 (Hit@3 = 100%).
- 약점은 **citation 1.20** — baseline 프롬프트가 출처 인용을 강제하지 않아서 답변에 `[source_id]` 가 거의 안 들어감. EXP-2 가 이 부분을 노립니다.
- **completeness 2.20** — 다중 문서 종합 질문에서 한 source 만 풀어쓰고 나머지는 누락하는 경향. EXP-1 의 hybrid retriever 가 보완 시도.


## 5. 성능 개선 실험 (Baseline vs EXP-1 vs EXP-2)

같은 질문셋 20문항에 세 가지 setup 을 돌리고 정량/정성 비교합니다.

| Setup | 변경한 것 | bsvibe 미러 |
|---|---|---|
| **Baseline** | (기준) vector top-5 + 평범한 RAG 프롬프트 | — |
| **EXP-1 (hybrid)** | retriever 만 hybrid (Vector + BM25 + Graph, RRF) | `backend/knowledge/retrieval/hybrid_search.py` |
| **EXP-2 (orchestrator)** | prompt 만 KnowledgeAnswerOrchestrator 패턴 | `backend/workflow/application/knowledge_orchestrator.py` |

평가:
- 검색 metric: Hit@1, Hit@3, Hit@5, MRR
- 답변 metric: LLM-as-judge (gpt-4o-mini, 4기준 0-3점)


### 5.1 종합 metric 비교

In [ ]:
import json
from pathlib import Path
ROOT = Path.cwd().resolve()
while ROOT.name != "metacode_study" and ROOT != ROOT.parent: ROOT = ROOT.parent
RES = ROOT / "assignments/week4-rag-eval/data/results"
baseline = json.load(open(RES / "baseline.json"))
exp1 = json.load(open(RES / "exp1_hybrid.json"))
exp2 = json.load(open(RES / "exp2_orchestrator.json"))
judged = json.load(open(RES / "judged.json"))

def retr_metrics(setup):
    rows = setup["results"]
    h1 = sum(1 for r in rows if r["retrieval_metrics"]["hit@1"]) / len(rows)
    h3 = sum(1 for r in rows if r["retrieval_metrics"]["hit@3"]) / len(rows)
    h5 = sum(1 for r in rows if r["retrieval_metrics"]["hit@5"]) / len(rows)
    ranks = [r["retrieval_metrics"]["first_match_rank"] for r in rows if r["retrieval_metrics"]["first_match_rank"]]
    mrr = sum(1/r for r in ranks) / len(rows)
    return h1, h3, h5, mrr

print("=== Retrieval 지표 ===")
b = retr_metrics(baseline); e1 = retr_metrics(exp1); e2 = retr_metrics(exp2)
print(f"{'metric':16}  {'baseline':>10}  {'exp1_hybrid':>13}  {'exp2_orch':>13}")
print(f"{'Hit@1':16}  {b[0]:10.1%}  {e1[0]:13.1%}  {e2[0]:13.1%}")
print(f"{'Hit@3':16}  {b[1]:10.1%}  {e1[1]:13.1%}  {e2[1]:13.1%}")
print(f"{'Hit@5':16}  {b[2]:10.1%}  {e1[2]:13.1%}  {e2[2]:13.1%}")
print(f"{'MRR':16}  {b[3]:10.3f}  {e1[3]:13.3f}  {e2[3]:13.3f}")
print(f"{'elapsed (s)':16}  {baseline['elapsed_s']:10.1f}  {exp1['elapsed_s']:13.1f}  {exp2['elapsed_s']:13.1f}")

print()
print("=== LLM-as-judge (gpt-4o-mini, 4기준 0-3점) ===")
def jp(j): return j["overall"]
bj, e1j, e2j = jp(judged['baseline']), jp(judged['exp1_hybrid']), jp(judged['exp2_orchestrator'])
print(f"{'criterion':16}  {'baseline':>10}  {'exp1_hybrid':>13}  {'exp2_orch':>13}")
for k, label in [("mean_faithfulness", "faithfulness"), ("mean_relevance", "relevance"),
                 ("mean_completeness", "completeness"), ("mean_citation", "citation"),
                 ("mean_total", "total /12"), ("mean_normalized", "normalized")]:
    print(f"{label:16}  {bj[k]:10.3f}  {e1j[k]:13.3f}  {e2j[k]:13.3f}")


=== Retrieval 지표 ===
metric            baseline      exp1_hybrid       exp2_orchestrator
Hit@1              95.0%         75.0%             95.0%
Hit@3             100.0%         95.0%            100.0%
Hit@5             100.0%         95.0%            100.0%
MRR                0.967         0.842             0.967
elapsed (s)         62.0          56.7              34.4

=== LLM-as-judge (gpt-4o-mini, 4기준 0-3점) ===
metric (LLM-judge)  baseline      exp1_hybrid       exp2_orchestrator
faithfulness        2.80/3        3.00/3              2.65/3
relevance           3.00/3        3.00/3              2.95/3
completeness        2.20/3        2.35/3              1.95/3
citation            1.20/3        1.45/3              1.50/3
total /12           9.20         9.80               9.05
normalized          0.767        0.817             0.754


**관찰**

1. **Retrieval 지표는 baseline 이 동률 또는 더 좋다** — Hit@1 baseline 95% vs exp1 75%. baseline 의 dense vector 검색이 이미 충분히 강함.
2. **그런데도 답변 quality 는 exp1 (hybrid) 이 최고** — normalized 0.817 vs baseline 0.767. **"retrieval metric ≠ answer quality"** 의 정량 증거.
3. **exp2 (orchestrator prompt) 는 citation 단독 최고** — 1.50 vs baseline 1.20. 프롬프트로 `[source_id]` 인용을 강제한 효과. 다만 completeness 가 낮아 (1.95) 종합점수는 baseline 보다 낮음.

해석 — exp1 의 hybrid 가 정답을 항상 1위로 못 가져오더라도 (3순위까지 도달 OK), Graph 와 BM25 가 보강하는 **인접 source** 가 LLM 의 답변 generation 에서 더 풍부한 grounding 을 제공 → completeness 향상.


### 5.2 질문별 breakdown (judge 정규화 점수)

질문 20개 각각의 normalized 점수 + Hit 여부. EXP-1 이 어떤 케이스에서 baseline 을 넘었는지.


In [ ]:
# per-question breakdown
per_q = {q['qid']: q for q in [json.loads(l) for l in (ROOT / 'assignments/week4-rag-eval/data/eval/questions.jsonl').read_text().splitlines() if l.strip()]}
by_b = {r['qid']: r for r in baseline['results']}
by_e1 = {r['qid']: r for r in exp1['results']}
by_e2 = {r['qid']: r for r in exp2['results']}
jb_b = {r['qid']: r for r in judged['baseline']['per_question']}
jb_e1 = {r['qid']: r for r in judged['exp1_hybrid']['per_question']}
jb_e2 = {r['qid']: r for r in judged['exp2_orchestrator']['per_question']}

print(f"{'qid':6}  {'difficulty':11}  {'b_hit1':>6}  {'e1_hit1':>7}  {'e2_hit1':>7}  {'b_norm':>6}  {'e1_norm':>7}  {'e2_norm':>7}  {'Δ(e1-b)':>8}")
for qid in sorted(by_b):
    diff = by_b[qid]['difficulty']
    bn, en, e2n = jb_b[qid]['normalized'], jb_e1[qid]['normalized'], jb_e2[qid]['normalized']
    print(f"{qid:6}  {diff:11}  {str(by_b[qid]['retrieval_metrics']['hit@1'])[:5]:>6}  {str(by_e1[qid]['retrieval_metrics']['hit@1'])[:5]:>7}  {str(by_e2[qid]['retrieval_metrics']['hit@1'])[:5]:>7}  {bn:6.3f}  {en:7.3f}  {e2n:7.3f}  {en-bn:+.3f}")


qid     difficulty   b_hit1  e1_hit1  e2_hit1  b_norm  e1_norm  e2_norm   Δ(e1-b)
q-001   simple_fact    True     True     True   0.750    1.000    0.667  +0.250
q-002   simple_fact    True     True     True   0.750    0.833    1.000  +0.083
q-003   multi_doc      True     True     True   0.750    0.833    0.750  +0.083
q-004   simple_fact    True     True     True   0.750    0.750    0.750  +0.000
q-005   simple_fact    True    False     True   0.583    0.667    0.750  +0.084
q-006   simple_fact    True     True     True   0.750    0.750    0.917  +0.000
q-007   multi_doc      True     True     True   0.833    0.833    0.667  +0.000
q-008   simple_fact    True    False     True   0.500    0.833    0.583  +0.333
q-009   simple_fact    True     True     True   0.750    0.750    0.667  +0.000
q-010   multi_doc      True    False     True   1.000    1.000    0.833  +0.000
q-011   multi_doc      True     True     True   1.000    0.833    0.667  -0.167
q-012   simple_fact   False    False  

### 5.3 EXP-1 (Hybrid retriever) — 결정적 사례

bsvibe `hybrid_search.py` 의 RRF (Reciprocal Rank Fusion) 로 Vector + BM25 + Graph 세 갈래를 결합. 가장 큰 향상이 있던 질문 + retrieval 단계의 차이 비교.


In [ ]:
# EXP-1 winners
wins = sorted([q for q in by_b if jb_e1[q]['normalized'] - jb_b[q]['normalized'] >= 0.25])
for qid in wins[:2]:
    b_row, e1_row = by_b[qid], by_e1[qid]
    print(f"=== [{qid}] {b_row['question']} ===")
    print(f"expected: {b_row['expected_source_ids']}\n")
    print("--- baseline retrieval (vector top-5) ---")
    for h in b_row['hits']:
        mark = " ✓" if h['source_id'] in b_row['expected_source_ids'] else ""
        print(f"  {h['rank']}. sim={h['similarity']:.3f}  {h['source_id']:18}{mark}  {h['title']}")
    print("\n--- exp1 hybrid retrieval (RRF) ---")
    for h in e1_row['hits']:
        mark = " ✓" if h['source_id'] in b_row['expected_source_ids'] else ""
        via = "+".join(h.get('matched_via', []))
        print(f"  {h['rank']}. RRF={h['similarity']:.4f}  via={via:<14}  {h['source_id']:18}{mark}  bm25={h.get('bm25_rank')} vec={h.get('vector_rank')} graph={h.get('graph_rank')}  {h['title']}")
    print("\n--- baseline answer ---\n ", b_row['answer'][:500])
    print("\n--- exp1 answer ---\n ", e1_row['answer'][:500])
    print(f"\n--- judge ---\n  baseline: {jb_b[qid]}\n  exp1    : {jb_e1[qid]}\n")


=== [q-001] BSVibe 모노레포는 어떤 구성으로 돌아가나요? ===
expected: ['bsvibe-src-001']

--- baseline retrieval (vector top-5) ---
  1. sim=0.607  bsvibe-src-001     ✓  
  2. sim=0.441  bsvibe-src-030      
  3. sim=0.419  bsvibe-src-004      
  4. sim=0.385  bsvibe-src-006      
  5. sim=0.381  bsvibe-src-031      

--- exp1 hybrid retrieval (RRF) ---
  1. RRF=0.0492  via=bm25+graph+vector  bsvibe-src-001     ✓  bm25=1 vec=1 graph=1  BSVibe monorepo layout and local stack
  2. RRF=0.0484  via=bm25+graph+vector  bsvibe-src-030      bm25=2 vec=2 graph=2  compose.prod.yaml overrides dev defaults for prod env-driven startup
  3. RRF=0.0476  via=bm25+graph+vector  bsvibe-src-004      bm25=3 vec=3 graph=3  Resolved decisions are retrieved from settled garden notes
  4. RRF=0.0308  via=bm25+vector     bsvibe-src-031      bm25=5 vec=5 graph=None  Deploy README documents the migrate-on-boot and single-migrator contract
  5. RRF=0.0156  via=bm25            bsvibe-vault-001    bm25=4 vec=None graph=None  Fact 

**해석 (EXP-1)** — winners 의 공통 패턴은 hybrid 가 baseline 보다 **다른 source_id 를 top-5 안에 더 채워 넣었다**는 점입니다. baseline 은 비슷한 청크들이 top 을 점유 (low diversity), hybrid 는 BM25 의 keyword 매칭 + Graph 의 1-hop 이웃이 보강해 inspired source 가 답변 generation 에 들어가 completeness 가 올라갑니다.


### 5.4 EXP-2 (Orchestrator prompt) — 인용 효과

bsvibe `knowledge_orchestrator.py` 의 system prompt + grounding-block + cap. baseline 과 동일한 vector retriever 를 쓰지만 **프롬프트만** 교체.


In [ ]:
# EXP-2 demonstration on a couple of picks
for qid in ["q-002", "q-018"]:
    b, e2 = by_b[qid], by_e2[qid]
    print(f"=== [{qid}] {b['question']} ===\n")
    print("--- baseline answer ---\n ", b['answer'][:480])
    print("\n--- exp2 orchestrator answer ---\n ", e2['answer'][:480])
    print(f"\n--- judge ---")
    print(f"  baseline    : citation={jb_b[qid]['citation']}  total={jb_b[qid]['total']}")
    print(f"  exp2 orchest: citation={jb_e2[qid]['citation']}  total={jb_e2[qid]['total']}\n")


=== [q-002] workspace 단위로 vault와 retriever를 어떻게 묶나요? ===

--- baseline answer ---
  workspace 단위로 vault와 retriever를 묶는 방법은 KnowledgeFactory가 있습니다.

KnowledgeFactory는 region과 workspace_id에 따라 every vault를 roots하고, workspace-scoped Vault와 GardenWriter objects를 exposes합니다. 또한, workspace-scoped retriever를 constructs하여 runtime components에 사용됩니다.

이러한 mechanism은 workspace scoping이 first-class RAG metadata requirement가 되도록 합니다.

KnowledgeFactory의 role을 요약하면 다음과 같습니다:

* region과 workspace_id에 따라 every vault를 roots합니다.
* workspace-scoped Vault와 GardenWriter objects를 

--- exp2 orchestrator answer ---
  workspace 단위로 vault와 retriever를 묶는 방법은 KnowledgeFactory가 제공하는 constructor입니다. KnowledgeFactory는 region/workspace_id에 every vault를 roots하고, workspace-scoped Vault와 GardenWriter objects를 exposes하며, runtime components에 사용되는 workspace-scoped retriever를 constructs합니다. [bsvibe-src-002]

--- judge ---
  baseline    : citation=1  total=9
  exp2 orchest: citation=3  total=12

=== [q-018] SettleWorker의 절차와

**해석 (EXP-2)** — citation 점수 향상 (1.20 → 1.50) 의 주된 원인은 system prompt 에 명시된 ``Cite knowledge with [source_id] in-line`` 한 줄. completeness 가 살짝 떨어진 것은 grounding 을 cap 5개로 자르고 per-statement 500 chars 로 clamp 하다보니 일부 긴 문서 (q-018 같이 두 파이프라인 비교) 에서 정보가 잘려나간 영향으로 보입니다.

→ orchestrator 패턴은 **citation 규율** 이 중요한 시나리오 (감사·근거 추적·법무) 에서 효과적이고, **다중 문서 종합** 이 중요한 시나리오에선 cap 을 풀거나 EXP-1 의 hybrid 와 결합하는 것이 옳습니다.


## 6. 결론

| 평가축 | 승자 | 비고 |
|---|---|---|
| Hit@1 | baseline (95.0%) | dense top-1 이 이미 강함 |
| Hit@3 | baseline (100%) = exp2 | top-3 안에서는 동률 |
| MRR | baseline (0.967) = exp2 | retriever 단순 비교에선 동률 |
| LLM-judge total | **exp1 hybrid (9.80/12)** | hybrid 가 종합 답변 quality 1위 |
| faithfulness | **exp1 hybrid (3.00)** | grounding 다양성 효과 |
| completeness | **exp1 hybrid (2.35)** | 다중 source 회수 효과 |
| citation | **exp2 orchestrator (1.50)** | 프롬프트의 인용 강제 |

**최종 권장 조합**: EXP-1 의 hybrid retriever + EXP-2 의 orchestrator prompt 동시 적용 (이번 과제 범위 밖, 다음 단계).

**핵심 교훈** — retrieval metric (Hit@k, MRR) 만 보면 baseline 이 더 좋아 보이지만, **답변 quality 는 retriever 의 다양성과 prompt 의 규율** 에 더 크게 좌우됩니다. RAG 성능 평가는 반드시 retrieval + answer 양쪽 metric 을 같이 봐야 합니다.
